### PR TITLE
[HOLD] "link_text" Feature

### PR DESCRIPTION
**Pull request by jefleponot**
Hello

I have write a new feature : "link_text".
The feature adds a new selector preconize in webdriver protocols.

SelfTests OK.
Documentations updated.

I'm a newbee in github . when I use the command : "git push origin linkText" - it is forbidden 

remote: Permission to n1k0/casperjs.git denied to jefleponot.
fatal: unable to access 'https://github.com/n1k0/casperjs.git/': The requested URL returned error: 403

So I send it to my clone repository. Is it the correct way ?

Please Help me.

### CODE REVIEW

**Comment by paazmaya** [comment]
Can you please create another pull request in which you have the TrifleJS related changes?

**Comment by jefleponot** [comment]
There is no TrifleJS changes in my PR. It was a mistake in my first push name.

**Comment by paazmaya** [comment]
Umm... have you taken a look at the "Files changed" tab?

**Comment by jefleponot** [comment]
I made a new fix for tabs :-)

**Comment by mickaelandrieu** [comment]
Hi.

I don't realy understand what is the purpose of this contribution. Can you explain me ?

Regards,

**Comment by jefleponot** [comment]
Hi Mickael,

Casper is a great tools for scripting but it miss features.
Css and xpath selector are usefuls but not very user-friendly so anothers shortcuts selectors have been inserted in all web testing frameworks :

**Comment by mickaelandrieu** [comment]
Hi, can you please squash your commits? I Will make a review on monday /c @paazmaya

**Comment by jefleponot** [comment]
Thank you,
I'll try to do it.
my first squash commits

File: .jshintrc [hunk 1]
```diff
@@ -3,11 +3,11 @@
     "browser":       true,
     "debug":         true,
     "devel":         true,
-    "eqeqeq":        true, 
+    "eqeqeq":        true,
     "evil":          true,
     "maxparams":     5,
     "maxdepth":      3,
-    "maxstatements": 15,
+    "maxstatements": 50,
```
Comment (mickaelandrieu) [inline .jshintrc]: nope.

You need to refactor your code to be compliant with our standards.

File: .travis.yml [hunk 2]
```diff
@@ -47,14 +47,20 @@ env:
       ENGINE_ARCHIVE_URL="https://phantomjs.googlecode.com/files/phantomjs-1.9.2-linux-x86_64.tar.bz2"
```
Comment (mickaelandrieu) [inline .travis.yml]: can you make a distinct PR for this update ? we will merge it before this one.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -29,7 +29,7 @@ Sample commit message:
 
 Run CasperJS' test suite to see you didn't break something:
 
-    $ casperjs selftest
+    $ casperjs selftest && casper selftest --engine=slimerjs
```
Comment (mickaelandrieu) [inline CONTRIBUTING.md]: I don't understand why do you need this here, in this PR I mean.

**Comment by mickaelandrieu** [comment]
There is too much things unrelated to your feature, you need to remove this stuff and rebase from master branch.

File: bin/bootstrap.js [hunk 4]
```diff
@@ -40,8 +40,17 @@ if ('process' in this && process.title === "node") {
 // phantom check
 if (!('phantom' in this)) {
     console.error('CasperJS needs to be executed in a PhantomJS environment http://phantomjs.org/');
+} else {
+    if (phantom.version.major === 2) {
```
Comment (mickaelandrieu) [inline bin/bootstrap.js]: this update is already available on master branch ...

File: bin/bootstrap.js [hunk 5]
```diff
@@ -87,18 +96,22 @@ CasperError.prototype = Object.getPrototypeOf(new Error());
         return;
     }
 
+    function __exit(statusCode){
```
Comment (mickaelandrieu) [inline bin/bootstrap.js]: same here.

**Comment by mickaelandrieu** [comment]
"Houston, we got a problem".

This is not good, you need to rebase master branch of casperjs official repository.
Assuming you are working on your fork, this is a common way to do:

```bash
// add a remote for n1k0/casperjs repository
git remote add upstream https://github.com/n1k0/casperjs.git

// then update your "local" master branch

git checkout master
git fetch -a
git rebase -i upstream/master (you will probably need to fix some bugs)
git push origin master

// then go on the "feature" branch
git checkout origin linkText
git rebase -i origin/master (you will probably need to fix some bugs)

// so at this point, your master branch is the same as our master branch and only the feature should be in your commits

git push -f origin linkText
```

from [Syncing a fork](https://help.github.com/articles/syncing-a-fork/)

Let me know when it's done (and don't give up, I want linkText() to be part of casperjs API :+1:  )

**Comment by jefleponot** [comment]
Hi Mickael,

Your are totaly right.
I will be my best today to do all you said.

Best Regards
JEF

**Comment by jefleponot** [comment]
Must I do merge with upstream/master to test rebase (integration) ?

File: modules/clientutils.js [hunk 6]
```diff
@@ -356,9 +423,11 @@
                 var pSelector = this.processSelector(selector);
                 if (pSelector.type === 'xpath') {
                     return this.getElementByXPath(pSelector.path, scope);
-                } else {
-                    return scope.querySelector(pSelector.path);
                 }
+                if (pSelector.type === 'link text') {
```
Comment (mickaelandrieu) [inline modules/clientutils.js]: maybe ``link_text`` instead of ``link text`` ?

File: tests/sample_modules/csmodule.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+/*global casper*/
```
Comment (mickaelandrieu) [inline tests/sample_modules/csmodule.js]: why do you need this file ?

File: tests/suites/casper/flow.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
+/*global casper*/
```
Comment (mickaelandrieu) [inline tests/suites/casper/flow.js]: I'm ok for add some tests, but they are not related to your feature.

File: tests/suites/casper/linktext.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+            type: 'link text',
+            path: 'AN UPPERCASE LINK TO THE FORM'
+        }, 'link text selector, without upercase style, can find an element');
+        test.assertExists({
+            type: 'link text',
+            path: 'a lowercase link to the form'
+        }, 'link text selector, with lowercase style, can find an element');
+	test.assertExists({
+            type: 'link text',
+            path: 'A Capitalize Link To The Form'
```
Comment (mickaelandrieu) [inline tests/suites/casper/linktext.js]: :+1:

File: tests/suites/casper/visible.js [hunk 10]
```diff
@@ -2,11 +2,11 @@
...
 casper.test.begin('visibility tests', 5, function(test) {
     casper.start('tests/site/visible.html', function() {
-        test.assert(!this.visible('#img1'), 'Casper.visible() can detect if an element is invisible');
+        test.assert(this.visible('#img1'), 'Casper.visible() can detect if an element is invisible');
         test.assert(this.visible('#img2'), 'Casper.visible() can detect if an element is visible');
         test.assert(!this.visible('#img3'), 'Casper.visible() can detect if an element is invisible');
         test.assert(this.visible('img'), 'Casper.visible() can detect if an element is visible');
-        this.waitWhileVisible('#img1', function() {
+        this.waitUntilVisible('#img1', function() {
             test.pass('Casper.waitWhileVisible() can wait while an element is visible');
```
Comment (mickaelandrieu) [inline tests/suites/casper/visible.js]: nope.
[waitUntilVisible](http://casperjs.readthedocs.org/en/latest/modules/casper.html?highlight=waituntilvisible#waituntilvisible) and [waitWhileVisible](http://casperjs.readthedocs.org/en/latest/modules/casper.html?highlight=waituntilvisible#waitwhilevisible) don't have the same behavior.

File: modules/clientutils.js (hunk 6 above)
Comment (jefleponot) [inline modules/clientutils.js]: i don't know
it's as you want
if you think it's better

File: tests/sample_modules/csmodule.js (hunk 7 above)
Comment (jefleponot) [inline tests/sample_modules/csmodule.js]: No need at all.
it's just my "jshint" test that forces to me that lines

File: tests/suites/casper/flow.js (hunk 8 above)
Comment (jefleponot) [inline tests/suites/casper/flow.js]: When I have used  "jshint" (thanks to Juga Paazmaya) a lot of errors appears.
So I try to do everything to pass them.

erasing blank chars, count functions, add comments ...

File: tests/suites/casper/linktext.js (hunk 9 above)
Comment (jefleponot) [inline tests/suites/casper/linktext.js]: :-)

File: tests/suites/casper/visible.js (hunk 10 above)
Comment (jefleponot) [inline tests/suites/casper/visible.js]: there is a bug code in the test command and it will be better to add anotther test

**Comment by jefleponot** [comment]
squash with visible test done.
selftest ok
jshint ok
push done

another note ?

**Comment by mickaelandrieu** [comment]
Yes, mono seems unreachable... I relaunch the build.

ping @n1k0 @paazmaya : can you take a look to this pull request ? for me it's :+1:

**Comment by jefleponot** [comment]
:+1:  French : C'est ma première INTÉGRATION!!! Hourra 
English: It's my first iNSERT Pull Request!!! hurray

**Comment by n1k0** [comment]
TBH I don't clearly see the added value of having this while we could simply rely on some simple xpath expression; that's what we do with `clickLabel`, and `//a[text()="blah"` seems to work just fine…

**Comment by jefleponot** [comment]
Hi nicolas,

Adding this feature is a quite important thing beacause :
- In html tags, texts are always written differently, using xpath pattern to match text is not obviusly, example :
<a>Hello</a>
<a>  Hello  </a>
<a>                Hello  </a>
<a><span>H<span><span>ello</span></a>
<a>  Hello 
</a>

witch pattern do you need : //a[text()="Hello"] OR //a[normalize-text(text())="Hello"] OR //a[contains(text(),"Hello")] OR //a[string()="Hello"] ... or a mixture ...
AND no one in all of them will works if css style is applied like "text-transform: Capitalize"

So an help is required : link text

Please just have a look to previous links.

Ps: why do you need a "ClickLabel" Method ? we have yet a click method with selector or xpath arg.

File: modules/casper.js [hunk 11]
```diff
@@ -705,7 +724,7 @@ Casper.prototype.evaluate = function evaluate(fn, context) {
         return utils.clone(this.page.evaluate(fn));
     } else if (arguments.length === 2) {
         // check for closure signature if it matches context
-        if (utils.isObject(context) && eval(fn).length === Object.keys(context).length) {
+        if (utils.isObject(context) && fn.length === Object.keys(context).length) {
```
Comment (paazmaya) [inline modules/casper.js]: Good catch, as the `fn` is documented to be a function, this should be safe

File: docs/modules/clientutils.rst [hunk 12]
```diff
@@ -214,6 +214,25 @@ Retrieves boundaries for all DOM element matching the provided :ref:`selector <s
...
+.. index:: Link text
+
+``getElementByLinkText()``
+-------------------------------------------------------------------------------
+
+**Signature:** ``getElementByLinkText(String text [, HTMLElement scope])``
+
+Retrieves a single DOM element matching a given :ref:`Link text <selectors>`.
+
+.. versionadded:: 1.0
```
Comment (paazmaya) [inline docs/modules/clientutils.rst]: Perhaps all there "versionadded" parts should have unified version number in them and the way it would seem to be now, it is `1.1`.

File: modules/clientutils.js [hunk 13]
```diff
@@ -54,12 +54,14 @@
             -1, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
             41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, -1, -1, -1, -1, -1
         );
-        var SUPPORTED_SELECTOR_TYPES = ['css', 'xpath'];
+        var SUPPORTED_SELECTOR_TYPES = ['css', 'xpath','link text'];
```
Comment (paazmaya) [inline modules/clientutils.js]: One space after comma

File: modules/clientutils.js (hunk 6 above)
Comment (paazmaya) [inline modules/clientutils.js]: I agree with using `link_text` instead of `link text`.

File: modules/clientutils.js (hunk 13 above)
Comment (jefleponot) [inline modules/clientutils.js]: ok
:-)

File: docs/modules/clientutils.rst (hunk 12 above)
Comment (jefleponot) [inline docs/modules/clientutils.rst]: ok

**Comment by jefleponot** [comment]
2 votes for "link_text" (paazmaya+mickaelandrieu)
1 vote for "link text"
0 vote for "linkText" 

:+1:

File: modules/clientutils.js (hunk 6 above)
Comment (jefleponot) [inline modules/clientutils.js]: 2 votes for "link_text" (paazmaya+mickaelandrieu)
1 vote for "link text"
0 vote for "linkText" 

:+1:

File: modules/clientutils.js [hunk 14]
```diff
@@ -467,6 +536,22 @@
...
 
         /**
+         * Retrieves information about the node computed style the provided selector.
+         *
+         * @param  String|Object  selector  CSS3/XPath selector
+         * @return Object
+         */
+        this.getElementStyle = function getElementStyle(selector) {
+            var element = this.findOne(selector);
+            var styles = window.getComputedStyle(element, null) ;
```
Comment (paazmaya) [inline modules/clientutils.js]: One extra space

File: modules/clientutils.js [hunk 15]
```diff
@@ -467,6 +536,22 @@
...
+            var properties = {};
+            [].forEach.call(styles,function(propertyName){
```
Comment (paazmaya) [inline modules/clientutils.js]: space after comma

Comment (jefleponot) [inline modules/clientutils.js]: done

File: modules/clientutils.js (hunk 14 above)
Comment (jefleponot) [inline modules/clientutils.js]: done

**Comment by jefleponot** [comment]
thank you @paazmaya

**Comment by paazmaya** [comment]
Great that you fixed many of the linting errors.
There seems to be plenty of whitespace inconsistencies, mainly between the next word and comma.
Can you please fix those?

**Comment by paazmaya** [comment]
Also the "versionadded" in all of the new methods

**Comment by jefleponot** [comment]
ok

**Comment by paazmaya** [comment]
Yes, in the documentation

**Comment by jefleponot** [comment]
done

**Comment by jefleponot** [comment]
COUCOU

**Comment by mickaelandrieu** [comment]
@jefleponot you need to rebase it.
@n1k0 so ... do you (finaly) agree on this or not ? If not, it's better to close this one instead of ask to  @jefleponot to work more on it...

**Comment by jefleponot** [comment]
Hi Mickael, what does that means  [HOLD ] ?

jef

**Comment by mickaelandrieu** [comment]
this means I need a decision from @n1k0.

File: modules/casper.js (hunk 11 above)
Comment (n1k0) [inline modules/casper.js]: PhantomJS'  `evaluate()` accepts a string containing a function declaration as a first arg. This part of the code is to ensure we supports this use case.

File: modules/clientutils.js [hunk 16]
```diff
@@ -761,7 +928,7 @@
          */
         this.scrollToBottom = function scrollToBottom() {
             this.scrollTo(0, this.getDocumentHeight());
-        },
+        };
```
Comment (n1k0) [inline modules/clientutils.js]: Good catch.

File: tests/sample_modules/csmodule.js (hunk 7 above)
Comment (n1k0) [inline tests/sample_modules/csmodule.js]: >  it's just my "jshint" test that forces to me that lines

So please remove it from this patch.

Comment (n1k0) [inline tests/sample_modules/csmodule.js]: Also, I still don't get how is this new file related to the current patch.

File: tests/suites/casper/flow.js [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+/*jshint strict:false*/
+// Generated by CoffeeScript 1.4.0
```
Comment (n1k0) [inline tests/suites/casper/flow.js]: Why would we want this file.

File: tests/suites/coffee.js [hunk 18]
```diff
@@ -0,0 +1,32 @@
+/*global casper*/
+/*jshint strict:false*/
+// Generated by CoffeeScript 1.4.0
```
Comment (n1k0) [inline tests/suites/coffee.js]: We don't want these files, please remove them from the patch.

File: tests/suites/utils.js [hunk 19]
```diff
@@ -96,7 +96,7 @@ if (utils.gteVersion(phantom.version, '1.9.0')) {
                           modifiers.ctrl | modifiers.alt | modifiers.shift,
                          'computeModifier() computes a "ctrl+alt+shift" modifier');
         test.assertThrows(utils.computeModifier, ["chucknorris", modifiers],
-                         'computeModifier() checks for a valid modifier');
+                         'computeModifier() checks for a valid modifier '+ JSON.stringify(modifiers,null,4));
```
Comment (n1k0) [inline tests/suites/utils.js]: This patch definitely does to match, please keep it focused on the feature you're actually trying to implement.

File: docs/modules/clientutils.rst [hunk 20]
```diff
@@ -214,6 +214,25 @@ Retrieves boundaries for all DOM element matching the provided :ref:`selector <s
...
 It returns an array of objects each having four keys: ``top``, ``left``, ``width`` and ``height``.
 
+.. index:: Link text
+
+``getElementByLinkText()``
+-------------------------------------------------------------------------------
+
+**Signature:** ``getElementByLinkText(String text [, HTMLElement scope])``
+
+Retrieves a single DOM element matching a given :ref:`Link text <selectors>`.
```
Comment (n1k0) [inline docs/modules/clientutils.rst]: This isn't what the code does.

File: docs/modules/clientutils.rst [hunk 21]
```diff
@@ -214,6 +214,25 @@ Retrieves boundaries for all DOM element matching the provided :ref:`selector <s
...
+
+.. versionadded:: 1.1
+
+The ``scope`` argument allow to set the context for searching text on "a" tags::
```
Comment (n1k0) [inline docs/modules/clientutils.rst]: Why only `a` tags? What about `<span onclick="fn()">text</span>`, `<button onclick…>` and friends?

File: docs/modules/clientutils.rst [hunk 22]
```diff
@@ -233,6 +252,21 @@ The ``scope`` argument allow to set the context for executing the XPath query::
...
     __utils__.getElementByXPath('.//a', __utils__.findOne('div.main'));
 
+.. index:: Link text
+
+``getElementsByLinkText()``
+-------------------------------------------------------------------------------
+
+**Signature:** ``getElementsByLinkText(String text [, HTMLElement scope])``
+
+Retrieves all DOM elements matching a given :ref:`link text <selectors>`, if any.
```
Comment (n1k0) [inline docs/modules/clientutils.rst]: Do we need that plural form version? Sample real-world use case please.

File: docs/modules/clientutils.rst [hunk 23]
```diff
@@ -233,6 +252,21 @@ The ``scope`` argument allow to set the context for executing the XPath query::
...
+
+.. versionadded:: 1.1.4
+
+The ``scope`` argument allows to set the context for searching text on "a" tags.
+
+.. _clientutils_getfieldvalue:
```
Comment (n1k0) [inline docs/modules/clientutils.rst]: How is this section related to this method?

File: docs/selectors.rst [hunk 24]
```diff
@@ -97,3 +97,26 @@ To ease the use and reading of XPath expressions, a ``selectXPath`` helper is av
 .. warning::
 
    The only limitation of XPath use in CasperJS is in the :ref:`casper.fill() <casper_fill>` method when you want to fill **file fields**; PhantomJS natively only allows the use of CSS3 selectors in its `uploadFile method <https://github.com/ariya/phantomjs/wiki/API-Reference#wiki-webpage-uploadFile>`_, hence this limitation.
+   
+.. index:: Link text
+
+Link text
+---------
+
+.. versionadded:: 1.1.4
```
Comment (n1k0) [inline docs/selectors.rst]: What is that version number refering to?

File: modules/casper.js [hunk 25]
```diff
@@ -2011,6 +2029,20 @@ Casper.prototype.visible = function visible(selector) {
...
 
 /**
+ * Checks if an element matching the provided DOM CSS3/XPath/LinkText selector hasProperty
+ * current page DOM by checking that offsetWidth and offsetHeight are
+ * both non-zero.
+ *
+ * @param  String  selector  A DOM CSS3/XPath/LinkText selector
+ * @return Boolean
+ */
+Casper.prototype.getElementStyle = function getElementStyle(selector) {
```
Comment (n1k0) [inline modules/casper.js]: Please add documentation for this.

File: modules/casper.js [hunk 26]
```diff
@@ -2306,10 +2338,33 @@ Casper.prototype.waitWhileSelector = function waitWhileSelector(selector, then,
...
+ * match property condition in the remote DOM to process a next step.
+ *
+ * @param  String            selector     A DOM CSS3/XPath/LinkText selector
+ * @param  Function          condition    The test condition Function with Object of StyleProperty
+ * @param  Function          then         The next step to perform (optional)
+ * @param  Function          onTimeout    A callback function to call on timeout (optional)
+ * @param  Number            timeout      The max amount of time to wait, in milliseconds (optional)
+ * @return Casper
+ */
+Casper.prototype.waitStyleProperties = function waitStyleProperties( selector, condition, then, onTimeout, timeout) {
```
Comment (n1k0) [inline modules/casper.js]: Please add documentation for this.

File: modules/clientutils.js [hunk 27]
```diff
@@ -148,6 +150,53 @@
         };
 
         /**
+        * Prepares a string for xpath expression with the condition [text()=].
+        *
+        * @param  String  string
+        * @return String
+        */
+        this.quoteXPathAttributeString = function quoteXPathAttributeString(string) {
```
Comment (n1k0) [inline modules/clientutils.js]: docs?

File: modules/clientutils.js [hunk 28]
```diff
@@ -148,6 +150,53 @@
...
+            }
+            return '"' + string + '"';
+        };
+
+        /**
+        * Formats a string with passed parameters. Ported from nodejs `util.format()`.
+        *
+        * @return String
+        */
+        this.format = function format(f) {
```
Comment (n1k0) [inline modules/clientutils.js]: Why did you add this? It doesn't seem to be used in this patch.

File: modules/clientutils.js [hunk 29]
```diff
@@ -168,7 +217,23 @@
...
+
+        /**
+         * Checks if a given DOM element as a style propertyValue in remove page.
+         *
+         * @param  Object   element  DOM element
+         * @param  String   propertyName the style property
+         * @param  String   value    the value expected
+         * @return Boolean
+         */
+        this.elementStyle = function elementStyle(elem, propertyName) {
```
Comment (n1k0) [inline modules/clientutils.js]: docs? also, we'd want tests for this.

File: modules/clientutils.js [hunk 30]
```diff
@@ -54,12 +54,14 @@
...
         );
-        var SUPPORTED_SELECTOR_TYPES = ['css', 'xpath'];
+        var SUPPORTED_SELECTOR_TYPES = ['css', 'xpath', 'link_text'];
 
         // public members
         this.options = options || {};
         this.options.scope = this.options.scope || document;
 
+		// hack jshint limitation statements
+		(function(){
```
Comment (n1k0) [inline modules/clientutils.js]: jshint provides rules to alter this limit. Use these instead.

**Comment by n1k0** [comment]
While I could see how this feature could be interested for, say, 1% of the population of casperjs users, I'm definitely uncomfortable landing this in its current shape. 

The patch does far too much, misses tests and docs. Most importantly, as I can't really project myself into maintaining all this when you possibly leave, so I'm not gonna land it.

**Comment by jefleponot** [comment]
ok, bye

**Comment by n1k0** [comment]
> ok, bye

Sounds like I took the right decision.

> FYI, the "link Text" feature via Casper (hacked) is used by more than 54 people in my team to oversee the government's web applications

How does this decision prevent you to keep using this feature within your team? You can always extend the casper prototype and keep using it, right?

**Comment by mickaelandrieu** [comment]
@jefleponot you can extends CasperJS without any hack: http://casperjs.readthedocs.org/en/latest/extending.html

I consider this topic closed.

Regards,

**Comment by jefleponot** [comment]
Thanks Mickael for all.

To N1ko:

Hi Nicolas,
I'm pretty disappointed, on the one hand I have my government that refuses any contribution of its officials to the community as it gets the job full contributor like yours. On the other hand, contributors themselves who do not even realize how their work is used in all their possibilities.

Your project has been corrected, adapted and merged with several others to create a very complex tool used for behavioral tests, the SLA, the défacing, supervision, transfer auditors ...

Your programmer vision like mine does not cover the full extent of the mink writers and that's why I do not insist.

Take Care,
